src/datachain/lib/dc/storage.py

```diff
@@ -79,40 +154,35 @@ def from_storage(
     cache = session.catalog.cache
     client_config = session.catalog.client_config
 
-    list_ds_name, list_uri, list_path, list_ds_exists = get_listing(
-        uri, session, update=update
-    )
+    uris = uri if isinstance(uri, (tuple, list)) else [uri]
 
-    # ds_name is None if object is a file, we don't want to use cache
-    # or do listing in that case - just read that single object
-    if not list_ds_name:
-        dc = from_values(
+    if not uris:
+        raise ValueError("No URIs provided")
+
+    dc = None
+    listed_uris = set()
+
+    for single_uri in uris:
+        update_for_uri = update
+        if update:
+            _, lst_uri, _ = parse_listing_uri(os.fsdecode(single_uri), client_config)
+            if lst_uri in listed_uris:
+                listed_uris.add(lst_uri)
+                update_for_uri = False
+
+        chain = _extract_chain(
+            uri=single_uri,
+            file_type=file_type,
+            cache=cache,
             session=session,
             settings=settings,
             in_memory=in_memory,
-            file=[get_file_info(list_uri, cache, client_config=client_config)],
+            recursive=recursive,
+            object_name=object_name,
+            update=update_for_uri,
+            client_config=client_config,
         )
-        dc.signals_schema = dc.signals_schema.mutate({f"{object_name}": file_type})
-        return dc
 
-    if update or not list_ds_exists:
-        # disable prefetch for listing, as it pre-downloads all files
-        (
-            from_records(
-                DataChain.DEFAULT_FILE_RECORD,
-                session=session,
-                settings=settings,
-                in_memory=in_memory,
-            )
-            .settings(prefetch=0)
-            .gen(
-                list_bucket(list_uri, cache, client_config=client_config),
-                output={f"{object_name}": File},
-            )
-            .save(list_ds_name, listing=True)
-        )
-
-    dc = from_dataset(list_ds_name, session=session, settings=settings)
-    dc.signals_schema = dc.signals_schema.mutate({f"{object_name}": file_type})
+        dc = dc.union(chain) if dc else chain
 
-    return ls(dc, list_path, recursive=recursive, object_name=object_name)
+    return dc  # type: ignore[return-value]
```

Comment:
why do we ingore types here?

Reply:
Because mypy assumes that the value can be None, but when the reality is it cannot be none as we have enough safeguards above.

Reply:
let's put assert a line above then please

`assert dc`

Reply:
I am not sure what assert will achieve or improve here tbh, is it only to correct mypy type detection?

If for providing feedback to user, you can see we raise value error if uri is empty, the only case when it can be None.